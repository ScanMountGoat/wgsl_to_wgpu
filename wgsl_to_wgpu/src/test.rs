use super::*;
use indoc::indoc;

// Tokenstreams can't be compared directly using PartialEq.
// Use pretty_print to normalize the formatting and compare strings.
// Use a colored diff output to make differences easier to see.
#[cfg(test)]
#[macro_export]
macro_rules! assert_tokens_eq {
    ($a:expr, $b:expr) => {
        pretty_assertions::assert_eq!(
            $crate::pretty_print_rustfmt($a),
            $crate::pretty_print_rustfmt($b)
        )
    };
}

#[cfg(test)]
#[macro_export]
macro_rules! assert_tokens_snapshot {
    ($output:expr) => {
        let mut settings = insta::Settings::new();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_omit_expression(true);
        settings.bind(|| {
            insta::assert_snapshot!($crate::pretty_print_rustfmt($output));
        });
    };
}

#[cfg(test)]
#[macro_export]
macro_rules! assert_rust_snapshot {
    ($output:expr) => {
        let mut settings = insta::Settings::new();
        settings.set_prepend_module_to_snapshot(false);
        settings.set_omit_expression(true);
        settings.bind(|| {
            insta::assert_snapshot!($output);
        });
    };
}

#[test]
fn create_shader_module_push_constants() {
    let source = indoc! {r#"
        var<push_constant> consts: vec4<f32>;

        @fragment
        fn fs_main() -> @location(0) vec4<f32> {
            return consts;
        }
    "#};

    let actual = create_shader_module(source, "shader.wgsl", WriteOptions::default()).unwrap();
    assert_rust_snapshot!(actual);
}

#[test]
fn create_shader_multiple_entries() {
    let source = indoc! {r#"
        @group(0) @binding(0) var<uniform> a: f32;
        @group(0) @binding(1) var<uniform> b: f32;
        @group(0) @binding(2) var<uniform> c: f32;
        @group(0) @binding(3) var<uniform> d: u32;
        @group(0) @binding(4) var<uniform> e: u32;
        @group(0) @binding(5) var<uniform> f: u32;
        @group(0) @binding(6) var<uniform> g: u32;
        @group(0) @binding(7) var<uniform> h: u32;
        @group(0) @binding(8) var<uniform> i: f64;

        fn inner() -> f32 {
            return d;
        }

        fn inner_double() -> f64 {
            return i;
        }

        @vertex
        fn vs_main() {
            {
                let x = b;
                let y = f;
            }

            let x = h;

            switch e {
                default: {
                    let y = e;
                    return;
                }
            }
        }

        @fragment
        fn fs_main()  {
            let z = e;

            loop {
                let z = c;
            }

            if true {
                let x = h;
                let y = g;
            }
        }

        @compute @workgroup_size(1, 1, 1)
        fn main() {
            let y = inner();
            let z = f;
            loop {
                let w = g;
                let x = h;
            }
        }
    "#};

    let actual = create_shader_module(source, "shader.wgsl", WriteOptions::default()).unwrap();
    assert_rust_snapshot!(actual);
}

#[test]
fn create_shader_module_multiple_outputs() {
    let source = indoc! {r#"
        struct Output {
            @location(0) col0: vec4<f32>,
            @builtin(frag_depth) depth: f32,
            @location(1) col1: vec4<f32>,
        };

        @fragment
        fn fs_multiple() -> Output {}
    "#};
    let actual = create_shader_module(source, "shader.wgsl", WriteOptions::default()).unwrap();
    assert_rust_snapshot!(actual);
}

#[test]
fn create_shader_modules_source() {
    let source = "@fragment fn main() {}";
    let actual =
        create_shader_modules(source, WriteOptions::default(), demangle_identity).unwrap();
    assert_rust_snapshot!(actual);
}

#[test]
fn create_shader_modules_source_rustfmt() {
    let source = "@fragment fn main() {}";
    let actual = create_shader_modules(
        source,
        WriteOptions {
            rustfmt: true,
            ..Default::default()
        },
        demangle_identity,
    )
    .unwrap();
    assert_rust_snapshot!(actual);
}

#[test]
fn create_shader_module_consecutive_bind_groups() {
    let source = indoc! {r#"
        struct A {
            f: vec4<f32>
        };
        @group(0) @binding(0) var<uniform> a: A;
        @group(1) @binding(0) var<uniform> b: f32;
        @group(2) @binding(0) var<uniform> c: vec4<f32>;
        @group(3) @binding(0) var<uniform> d: mat4x4<f32>;

        @vertex
        fn vs_main() {}

        @fragment
        fn fs_main() {}
    "#};

    create_shader_module(source, "shader.wgsl", WriteOptions::default()).unwrap();
}

#[test]
fn create_shader_module_non_consecutive_bind_groups() {
    let source = indoc! {r#"
        @group(0) @binding(0) var<uniform> a: vec4<f32>;
        @group(1) @binding(0) var<uniform> b: vec4<f32>;
        @group(3) @binding(0) var<uniform> c: vec4<f32>;

        @fragment
        fn main() {}
    "#};

    let result = create_shader_module(source, "shader.wgsl", WriteOptions::default());
    assert!(matches!(
        result,
        Err(CreateModuleError::NonConsecutiveBindGroups)
    ));
}

#[test]
fn create_shader_module_repeated_bindings() {
    let source = indoc! {r#"
        struct A {
            f: vec4<f32>
        };
        @group(0) @binding(2) var<uniform> a: A;
        @group(0) @binding(2) var<uniform> b: A;

        @fragment
        fn main() {}
    "#};

    let result = create_shader_module(source, "shader.wgsl", WriteOptions::default());
    assert!(matches!(
        result,
        Err(CreateModuleError::DuplicateBinding { binding: 2 })
    ));
}

fn items_to_tokens(items: Vec<(TypePath, TokenStream)>) -> TokenStream {
    let mut root = Module::default();
    root.add_module_items(&items, &ModulePath::default());
    root.to_tokens()
}

#[test]
fn write_vertex_module_empty() {
    let source = indoc! {r#"
        @vertex
        fn main() {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = vertex_struct_methods(&module, demangle_identity);

    assert_tokens_eq!(quote!(), items_to_tokens(actual));
}

#[test]
fn write_vertex_module_single_input_float32() {
    let source = indoc! {r#"
        struct VertexInput0 {
            @location(0) a: f32,
            @location(1) b: vec2<f32>,
            @location(2) c: vec3<f32>,
            @location(3) d: vec4<f32>,
        };

        @vertex
        fn main(in0: VertexInput0) {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = vertex_struct_methods(&module, demangle_identity);

    assert_tokens_snapshot!(items_to_tokens(actual));
}

#[test]
fn write_vertex_module_single_input_float64() {
    let source = indoc! {r#"
        struct VertexInput0 {
            @location(0) a: f64,
            @location(1) b: vec2<f64>,
            @location(2) c: vec3<f64>,
            @location(3) d: vec4<f64>,
        };

        @vertex
        fn main(in0: VertexInput0) {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = vertex_struct_methods(&module, demangle_identity);

    assert_tokens_snapshot!(items_to_tokens(actual));
}

#[test]
fn write_vertex_module_single_input_float16() {
    let source = indoc! {r#"
        enable f16;

        struct VertexInput0 {
            @location(0) a: f16,
            @location(1) b: vec2<f16>,
            @location(2) c: vec4<f16>,
        };

        @vertex
        fn main(in0: VertexInput0) {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = vertex_struct_methods(&module, demangle_identity);

    assert_tokens_snapshot!(items_to_tokens(actual));
}

#[test]
fn write_vertex_module_single_input_sint32() {
    let source = indoc! {r#"
        struct VertexInput0 {
            @location(0) a: i32,
            @location(1) b: vec2<i32>,
            @location(2) c: vec3<i32>,
            @location(3) d: vec4<i32>,

        };

        @vertex
        fn main(in0: VertexInput0) {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = vertex_struct_methods(&module, demangle_identity);

    assert_tokens_snapshot!(items_to_tokens(actual));
}

#[test]
fn write_vertex_module_single_input_uint32() {
    let source = indoc! {r#"
        struct VertexInput0 {
            @location(0) a: u32,
            @location(1) b: vec2<u32>,
            @location(2) c: vec3<u32>,
            @location(3) d: vec4<u32>,
        };

        @vertex
        fn main(in0: VertexInput0) {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = vertex_struct_methods(&module, demangle_identity);

    assert_tokens_snapshot!(items_to_tokens(actual));
}

#[test]
fn write_compute_module_empty() {
    let source = indoc! {r#"
        @vertex
        fn main() {}
    "#};

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = compute_module(&module, demangle_identity);

    assert_tokens_eq!(quote!(), actual);
}

#[test]
fn write_compute_module_multiple_entries() {
    let source = indoc! {r#"
        @compute
        @workgroup_size(1,2,3)
        fn main1() {}

        @compute
        @workgroup_size(256)
        fn main2() {}
    "#
    };

    let module = naga::front::wgsl::parse_str(source).unwrap();
    let actual = compute_module(&module, demangle_identity);

    assert_tokens_snapshot!(actual);
}

#[test]
fn create_shader_module_parse_error() {
    let source = indoc! {r#"
        var<push_constant> consts: vec4<f32>;

        @fragment
        fn fs_main() }
    "#};

    let result = create_shader_module(source, "shader.wgsl", WriteOptions::default());

    assert!(matches!(result, Err(CreateModuleError::ParseError { .. })));
}

#[test]
fn create_shader_module_semantic_error() {
    let source = indoc! {r#"
        var<push_constant> consts: vec4<f32>;

        @fragment
        fn fs_main() {
            consts.x = 1;
        }
    "#};

    let result = create_shader_module(
        source,
        "shader.wgsl",
        WriteOptions {
            validate: Some(Default::default()),
            ..Default::default()
        },
    );

    assert!(matches!(
        result,
        Err(CreateModuleError::ValidationError { .. })
    ));
}

fn demangle_underscore(name: &str) -> TypePath {
    // Preprocessors that support modules mangle absolute paths.
    // Use a very basic mangling scheme that assumes no '_' in the identifier name.
    // This allows testing the module logic without needing extra dependencies.
    // a_b_C -> a::b::C
    let components: Vec<_> = name.split("_").collect();
    let (name, parents) = components.split_last().unwrap();
    TypePath {
        parent: ModulePath {
            components: parents.iter().map(|p| p.to_string()).collect(),
        },
        name: name.to_string(),
    }
}

#[test]
fn single_root_module() {
    let output = create_shader_modules(
        include_str!("data/modules.wgsl"),
        WriteOptions {
            rustfmt: true,
            ..Default::default()
        },
        demangle_underscore,
    )
    .unwrap();

    assert_rust_snapshot!(output);
}

#[test]
fn add_single_root_module() {
    let mut root = Module::default();
    let options = WriteOptions {
        rustfmt: true,
        ..Default::default()
    };
    root.add_shader_module(
        include_str!("data/modules.wgsl"),
        None,
        options,
        ModulePath::default(),
        demangle_underscore,
    )
    .unwrap();

    let output = root.to_generated_bindings(options);
    assert_rust_snapshot!(output);
}

#[test]
fn add_duplicate_module_different_paths() {
    // Test shared types and handling of duplicate names.
    let mut root = Module::default();
    let options = WriteOptions {
        rustfmt: true,
        ..Default::default()
    };
    root.add_shader_module(
        include_str!("data/modules.wgsl"),
        None,
        options,
        ModulePath {
            components: vec!["shader1".to_string()],
        },
        demangle_underscore,
    )
    .unwrap();
    root.add_shader_module(
        include_str!("data/modules.wgsl"),
        None,
        options,
        ModulePath {
            components: vec!["shaders".to_string(), "shader2".to_string()],
        },
        demangle_underscore,
    )
    .unwrap();

    let output = root.to_generated_bindings(options);
    assert_rust_snapshot!(output);
}

#[test]
fn vertex_entries() {
    // Check vertex entry points and builtin attribute handling.
    let actual = create_shader_module(
        include_str!("data/vertex_entries.wgsl"),
        "shader.wgsl",
        WriteOptions {
            rustfmt: true,
            ..Default::default()
        },
    )
    .unwrap();

    assert_rust_snapshot!(actual);
}

#[test]
fn shader_stage_collection() {
    // Check the visibility: wgpu::ShaderStages::COMPUTE
    let actual = create_shader_module(
        include_str!("data/shader_stage_collection.wgsl"),
        "shader.wgsl",
        WriteOptions {
            rustfmt: true,
            derive_encase_host_shareable: true,
            ..Default::default()
        },
    )
    .unwrap();

    assert_rust_snapshot!(actual);
}
