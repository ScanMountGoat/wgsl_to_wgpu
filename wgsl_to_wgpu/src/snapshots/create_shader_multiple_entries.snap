---
source: wgsl_to_wgpu/src/test.rs
---
pub mod bind_groups {
    #[derive(Debug)]
    pub struct BindGroup0(wgpu::BindGroup);
    #[derive(Debug)]
    pub struct BindGroupLayout0<'a> {
        pub a: wgpu::BufferBinding<'a>,
        pub b: wgpu::BufferBinding<'a>,
        pub c: wgpu::BufferBinding<'a>,
        pub d: wgpu::BufferBinding<'a>,
        pub e: wgpu::BufferBinding<'a>,
        pub f: wgpu::BufferBinding<'a>,
        pub g: wgpu::BufferBinding<'a>,
        pub h: wgpu::BufferBinding<'a>,
        pub i: wgpu::BufferBinding<'a>,
    }
    const LAYOUT_DESCRIPTOR0: wgpu::BindGroupLayoutDescriptor = wgpu::BindGroupLayoutDescriptor {
        label: Some("LayoutDescriptor0"),
        entries: &[
            wgpu::BindGroupLayoutEntry {
                binding: 0,
                visibility: wgpu::ShaderStages::NONE,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 1,
                visibility: wgpu::ShaderStages::VERTEX,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 2,
                visibility: wgpu::ShaderStages::FRAGMENT,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 3,
                visibility: wgpu::ShaderStages::COMPUTE,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 4,
                visibility: wgpu::ShaderStages::VERTEX_FRAGMENT,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 5,
                visibility: wgpu::ShaderStages::VERTEX
                    .union(wgpu::ShaderStages::COMPUTE),
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 6,
                visibility: wgpu::ShaderStages::FRAGMENT
                    .union(wgpu::ShaderStages::COMPUTE),
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 7,
                visibility: wgpu::ShaderStages::VERTEX
                    .union(wgpu::ShaderStages::FRAGMENT)
                    .union(wgpu::ShaderStages::COMPUTE),
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
            wgpu::BindGroupLayoutEntry {
                binding: 8,
                visibility: wgpu::ShaderStages::NONE,
                ty: wgpu::BindingType::Buffer {
                    ty: wgpu::BufferBindingType::Uniform,
                    has_dynamic_offset: false,
                    min_binding_size: None,
                },
                count: None,
            },
        ],
    };
    impl BindGroup0 {
        pub fn get_bind_group_layout(device: &wgpu::Device) -> wgpu::BindGroupLayout {
            device.create_bind_group_layout(&LAYOUT_DESCRIPTOR0)
        }
        pub fn from_bindings(device: &wgpu::Device, bindings: BindGroupLayout0) -> Self {
            let bind_group_layout = device.create_bind_group_layout(&LAYOUT_DESCRIPTOR0);
            let bind_group = device
                .create_bind_group(
                    &wgpu::BindGroupDescriptor {
                        layout: &bind_group_layout,
                        entries: &[
                            wgpu::BindGroupEntry {
                                binding: 0,
                                resource: wgpu::BindingResource::Buffer(bindings.a),
                            },
                            wgpu::BindGroupEntry {
                                binding: 1,
                                resource: wgpu::BindingResource::Buffer(bindings.b),
                            },
                            wgpu::BindGroupEntry {
                                binding: 2,
                                resource: wgpu::BindingResource::Buffer(bindings.c),
                            },
                            wgpu::BindGroupEntry {
                                binding: 3,
                                resource: wgpu::BindingResource::Buffer(bindings.d),
                            },
                            wgpu::BindGroupEntry {
                                binding: 4,
                                resource: wgpu::BindingResource::Buffer(bindings.e),
                            },
                            wgpu::BindGroupEntry {
                                binding: 5,
                                resource: wgpu::BindingResource::Buffer(bindings.f),
                            },
                            wgpu::BindGroupEntry {
                                binding: 6,
                                resource: wgpu::BindingResource::Buffer(bindings.g),
                            },
                            wgpu::BindGroupEntry {
                                binding: 7,
                                resource: wgpu::BindingResource::Buffer(bindings.h),
                            },
                            wgpu::BindGroupEntry {
                                binding: 8,
                                resource: wgpu::BindingResource::Buffer(bindings.i),
                            },
                        ],
                        label: Some("BindGroup0"),
                    },
                );
            Self(bind_group)
        }
        pub fn set<P: SetBindGroup>(&self, pass: &mut P) {
            pass.set_bind_group(0, &self.0, &[]);
        }
    }
    #[derive(Debug, Copy, Clone)]
    pub struct BindGroups<'a> {
        pub bind_group0: &'a BindGroup0,
    }
    impl BindGroups<'_> {
        pub fn set<P: SetBindGroup>(&self, pass: &mut P) {
            self.bind_group0.set(pass);
        }
    }
    pub trait SetBindGroup {
        fn set_bind_group(
            &mut self,
            index: u32,
            bind_group: &wgpu::BindGroup,
            offsets: &[wgpu::DynamicOffset],
        );
    }
    impl SetBindGroup for wgpu::ComputePass<'_> {
        fn set_bind_group(
            &mut self,
            index: u32,
            bind_group: &wgpu::BindGroup,
            offsets: &[wgpu::DynamicOffset],
        ) {
            self.set_bind_group(index, bind_group, offsets);
        }
    }
    impl SetBindGroup for wgpu::RenderPass<'_> {
        fn set_bind_group(
            &mut self,
            index: u32,
            bind_group: &wgpu::BindGroup,
            offsets: &[wgpu::DynamicOffset],
        ) {
            self.set_bind_group(index, bind_group, offsets);
        }
    }
    impl SetBindGroup for wgpu::RenderBundleEncoder<'_> {
        fn set_bind_group(
            &mut self,
            index: u32,
            bind_group: &wgpu::BindGroup,
            offsets: &[wgpu::DynamicOffset],
        ) {
            self.set_bind_group(index, bind_group, offsets);
        }
    }
}
pub fn set_bind_groups<P: bind_groups::SetBindGroup>(
    pass: &mut P,
    bind_group0: &bind_groups::BindGroup0,
) {
    bind_group0.set(pass);
}
pub mod compute {
    pub const MAIN_WORKGROUP_SIZE: [u32; 3] = [1, 1, 1];
    pub fn create_main_pipeline(device: &wgpu::Device) -> wgpu::ComputePipeline {
        let module = super::create_shader_module(device);
        let layout = super::create_pipeline_layout(device);
        device
            .create_compute_pipeline(
                &wgpu::ComputePipelineDescriptor {
                    label: Some("Compute Pipeline main"),
                    layout: Some(&layout),
                    module: &module,
                    entry_point: Some("main"),
                    compilation_options: Default::default(),
                    cache: Default::default(),
                },
            )
    }
    pub const MAIN2_WORKGROUP_SIZE: [u32; 3] = [256, 1, 1];
    pub fn create_main2_pipeline(device: &wgpu::Device) -> wgpu::ComputePipeline {
        let module = super::create_shader_module(device);
        let layout = super::create_pipeline_layout(device);
        device
            .create_compute_pipeline(
                &wgpu::ComputePipelineDescriptor {
                    label: Some("Compute Pipeline main2"),
                    layout: Some(&layout),
                    module: &module,
                    entry_point: Some("main2"),
                    compilation_options: Default::default(),
                    cache: Default::default(),
                },
            )
    }
}
pub const ENTRY_VS_MAIN: &str = "vs_main";
pub const ENTRY_FS_MAIN: &str = "fs_main";
pub const ENTRY_MAIN: &str = "main";
pub const ENTRY_MAIN2: &str = "main2";
#[derive(Debug)]
pub struct VertexEntry<const N: usize> {
    pub entry_point: &'static str,
    pub buffers: [wgpu::VertexBufferLayout<'static>; N],
    pub constants: Vec<(&'static str, f64)>,
}
pub fn vertex_state<'a, const N: usize>(
    module: &'a wgpu::ShaderModule,
    entry: &'a VertexEntry<N>,
) -> wgpu::VertexState<'a> {
    wgpu::VertexState {
        module,
        entry_point: Some(entry.entry_point),
        buffers: &entry.buffers,
        compilation_options: wgpu::PipelineCompilationOptions {
            constants: &entry.constants,
            ..Default::default()
        },
    }
}
pub fn vs_main_entry() -> VertexEntry<0> {
    VertexEntry {
        entry_point: ENTRY_VS_MAIN,
        buffers: [],
        constants: Default::default(),
    }
}
#[derive(Debug)]
pub struct FragmentEntry<const N: usize> {
    pub entry_point: &'static str,
    pub targets: [Option<wgpu::ColorTargetState>; N],
    pub constants: Vec<(&'static str, f64)>,
}
pub fn fragment_state<'a, const N: usize>(
    module: &'a wgpu::ShaderModule,
    entry: &'a FragmentEntry<N>,
) -> wgpu::FragmentState<'a> {
    wgpu::FragmentState {
        module,
        entry_point: Some(entry.entry_point),
        targets: &entry.targets,
        compilation_options: wgpu::PipelineCompilationOptions {
            constants: &entry.constants,
            ..Default::default()
        },
    }
}
pub fn fs_main_entry(targets: [Option<wgpu::ColorTargetState>; 0]) -> FragmentEntry<0> {
    FragmentEntry {
        entry_point: ENTRY_FS_MAIN,
        targets,
        constants: Default::default(),
    }
}
pub const SOURCE: &str = include_str!("shader.wgsl");
pub fn create_shader_module(device: &wgpu::Device) -> wgpu::ShaderModule {
    let source = std::borrow::Cow::Borrowed(SOURCE);
    device
        .create_shader_module(wgpu::ShaderModuleDescriptor {
            label: None,
            source: wgpu::ShaderSource::Wgsl(source),
        })
}
pub fn create_pipeline_layout(device: &wgpu::Device) -> wgpu::PipelineLayout {
    device
        .create_pipeline_layout(
            &wgpu::PipelineLayoutDescriptor {
                label: None,
                bind_group_layouts: &[
                    &bind_groups::BindGroup0::get_bind_group_layout(device),
                ],
                push_constant_ranges: &[],
            },
        )
}
